evm: improve contracts (version 2)

gator-boi opened the pull request:
The purpose of this PR is to improve the Token Bridge Relayer contract. The following changes were made:
1) Added contract version.
2) Outbound transfers can be paused by the `owner`.
3) All fees and swapped tokens are sent to a designated `feeRecipient`. The `owner` can update this public key. 
4) Relayer fees and token swap rates can be changed by the `ownerAssistant`. The `owner` can update this public key.
5) Swap rates can now be updated in batches. 
6) Removed redundant constructor arguments.

gator-boi committed:
evm: simplify deployment

gator-boi committed:
evm: add version

gator-boi committed:
evm: add deployed addresses to testnet envs

gator-boi committed:
evm: add fee recipient and only owner method to update fee recipient

gator-boi committed:
evm: add test setup function

gator-boi committed:
evm: add ownerAssistant state variable

gator-boi committed:
evm: add batching to updateSwapRate

gator-boi committed:
evm: make outbound transfers pausable

gator-boi committed:
evm: remove deployed testnet addresses

SEJeff commented on evm/forge-scripts/deploy_contracts.sol:
```diff
@@ -42,6 +44,8 @@ contract ContractScript is Script {
 
         // verify getters
         require(relayer.chainId() == wormhole.chainId());
+        require(relayer.feeRecipient() == feeRecipient);
```
Maybe?

```suggestion
        require(relayer.feeRecipient() == feeRecipient);
        require(feeRecipient != address(0), "Fee recipient must be specified");
```

gator-boi committed:
evm: fix typos

gator-boi committed:
evm: add RELEASE_FEE_RECIPIENT to env files

gator-boi committed:
evm: update deployment scripts to handle batching

scnale committed:
evm: Adds overrides for configuration txs in Polygon and Fantom.

This avoids some gas pricing issues due to these networks not
implementing EIP1559 fully and instead only implementing an API
compatibility layer.
Additionally, gas limits are specified for polygon due to bad gas
estimations we've had from public endpoints.

scnale committed:
evm: adds a length factor in batch updates to gas limit in polygon.

SEJeff commented on evm/src/token-bridge-relayer/TokenBridgeRelayer.sol:
```diff
@@ -19,26 +20,34 @@ import "./TokenBridgeRelayerMessages.sol";
     using BytesLib for bytes;
 
+    // contract version
+    string public constant VERSION = "0.2.0";
```
Will this be updated for major changes? Or for every deployment?

gator-boi replied:
I think it should be updated for every deployment (at least the minor version number).

SEJeff commented on evm/src/token-bridge-relayer/TokenBridgeRelayerGovernance.sol:
```diff
@@ -166,26 +211,42 @@ abstract contract TokenBridgeRelayerGovernance is TokenBridgeRelayerGetters {
+
+            setSwapRate(swapRateUpdate[i].token, swapRateUpdate[i].value);
+
+            unchecked { i += 1; }
```
fancy fancy gas saver

SEJeff commented on evm/forge-test/TokenBridgeRelayer.t.sol:
```diff
@@ -666,6 +699,11 @@ contract TokenBridgeRelayerTest is Helpers, ForgeHelpers, Test {
             5 * avaxRelayer.relayerFeePrecision()
         );
 
+        // Pause and unpause the contract to confirm the contract allows
+        // outbound transfers after pause is disabled.
```
Smart
![think](https://media.giphy.com/media/d3mlE7uhX8KFgEmY/giphy.gif)

SEJeff commented on evm/forge-test/TokenBridgeRelayer.t.sol:
```diff
@@ -1298,65 +1425,25 @@ contract TokenBridgeRelayerTest is Helpers, ForgeHelpers, Test {
-            );
-        }
+        // make some assumptions about the fuzz test values
+        vm.assume(
+            normAmounts.transferAmount > 0 &&
+            amount < type(uint96).max
+        );
+        vm.assume(
+            normAmounts.toNative > 0 &&
+            toNativeTokenAmount < type(uint96).max &&
+            normAmounts.transferAmount > normAmounts.toNative + normAmounts.relayerFee
+        );
```
[vm.assume()](https://book.getfoundry.sh/cheatcodes/assume) actually reverts the test. You might consider constraining these numbers using [bound](https://book.getfoundry.sh/reference/forge-std/bound) where possible from forge-std as it limits the value mathematically instead of completely throwing away that full test run like vm.assume does.

An example might look like:
```solidity
toNativeTokenAmount = bound(toNativeTokenAmount, 0, type(uint96).max - 1)
```

SEJeff commented on evm/forge-test/TokenBridgeRelayer.t.sol:
```diff
@@ -1518,65 +1628,25 @@ contract TokenBridgeRelayerTest is Helpers, ForgeHelpers, Test {
+            normAmounts.toNative > 0 &&
+            toNativeTokenAmount < type(uint96).max &&
+            normAmounts.transferAmount > normAmounts.toNative + normAmounts.relayerFee
+        );
```
Same as previously. You can likely wrap some of these conditionals with [bound](https://book.getfoundry.sh/reference/forge-std/bound) for running the tests faster.

SEJeff commented on evm/forge-test/TokenBridgeRelayer.t.sol:
```diff
@@ -1893,57 +1987,17 @@ contract TokenBridgeRelayerTest is Helpers, ForgeHelpers, Test {
-            vm.assume(additionalGas > 0 && additionalGas < type(uint64).max);
-        }
+        // make test assumptions
+        vm.assume(additionalGas > 0 && additionalGas < type(uint64).max);
```
```solidity
additionalGas = bound(additionalGas, 1, type(uint64).max) - 1);
```

SEJeff commented on evm/forge-test/TokenBridgeRelayer.t.sol:
```diff
@@ -2102,58 +2179,25 @@ contract TokenBridgeRelayerTest is Helpers, ForgeHelpers, Test {
+            normAmounts.toNative > 0 &&
+            toNativeTokenAmount < type(uint96).max &&
+            normAmounts.transferAmount > normAmounts.toNative + normAmounts.relayerFee
+        );
```
And again. Look into using `bound()`

SEJeff reviewed (approved):
Some nits but nothing of substance. Nice work!

gator-boi committed:
Merge branch 'main' into evm/v2
